File: python/realtimex_sdk/v1/overrides/v1_workspace_streaming.py
# @manual-override — this file is never overwritten by generate-v1-sdk.mjs
"""
SSE streaming implementation for workspace chat.

The generated stub in v1_workspace.py returns a raw httpx Response;
use stream_workspace_chat() here for a typed AsyncIterator instead.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import AsyncIterator, Dict, Any, List, Optional

from ..client import DeveloperApiClient


@dataclass
class WorkspaceStreamSource:
    id: str
    url: Optional[str] = None
    title: Optional[str] = None
    score: Optional[float] = None


@dataclass
class WorkspaceStreamChunk:
    """A single SSE chunk from a workspace stream-chat response."""
    text_response: str
    id: str
    sources: List[WorkspaceStreamSource] = field(default_factory=list)
    close: bool = False
    error: Optional[str] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "WorkspaceStreamChunk":
        sources = [
            WorkspaceStreamSource(
                id=s.get("id", ""),
                url=s.get("url"),
                title=s.get("title"),
                score=s.get("score"),
            )
            for s in data.get("sources", [])
        ]
        return cls(
            text_response=data.get("textResponse", ""),
            id=data.get("id", ""),
            sources=sources,
            close=data.get("close", False),
            error=data.get("error"),
        )


async def stream_workspace_chat(
    client: DeveloperApiClient,
    slug: str,
    body: Optional[Dict[str, Any]] = None,
) -> AsyncIterator[WorkspaceStreamChunk]:
    """
    Execute a streamable chat with a workspace and yield typed SSE chunks.

    Example::

        sdk = RealtimeXSDK(realtimex={"api_key": "sk-..."})
        async for chunk in stream_workspace_chat(sdk.v1._client, "my-workspace", {"message": "Hello"}):
            print(chunk.text_response, end="", flush=True)
    """
    async with client.stream("POST", f"/v1/workspace/{slug}/stream-chat", body) as response:
        buffer = ""
        is_error_event = False

        async for raw in response.aiter_text():
            buffer += raw
            lines = buffer.split("\n")
            buffer = lines.pop()

            for line in lines:
                if line.startswith("event: error"):
                    is_error_event = True
                    continue
                if line.startswith("data: "):
                    json_str = line[6:].strip()
                    if json_str == "[DONE]":
                        is_error_event = False
                        continue
                    try:
                        data = json.loads(json_str)
                    except json.JSONDecodeError:
                        continue
                    chunk = WorkspaceStreamChunk.from_dict(data)
                    if is_error_event:
                        raise RuntimeError(chunk.error or "Unknown streaming error from server")
                    yield chunk

File: typescript/src/v1/overrides/v1WorkspaceStreaming.ts
// @manual-override — this file is never overwritten by generate-v1-sdk.mjs
// SSE streaming implementation for workspace chat.
// The generated stub in modules/v1Workspace.ts returns a raw Response;
// use streamChat() from this module for a typed AsyncGenerator instead.

import { DeveloperApiClient } from '../client';

export interface WorkspaceStreamChunk {
    /** The text fragment emitted by this SSE event */
    textResponse: string;
    /** Unique identifier for this chat session */
    id: string;
    /** Human-readable source references used for this response, if any */
    sources: Array<{ id: string; url?: string; title?: string; score?: number }>;
    /** Whether this is the final chunk in the stream */
    close: boolean;
    /** Error message, present only when the server emits an error event */
    error: string | null;
}

/**
 * Execute a streamable chat with a workspace and yield typed SSE chunks.
 *
 * @example
 * ```ts
 * const sdk = new RealtimeXSDK({ realtimex: { apiKey: 'sk-...' } });
 * for await (const chunk of streamWorkspaceChat(sdk.v1!._client, 'my-workspace', { message: 'Hello' })) {
 *   process.stdout.write(chunk.textResponse);
 * }
 * ```
 */
export async function* streamWorkspaceChat(
    client: DeveloperApiClient,
    slug: string,
    body?: Record<string, unknown>,
): AsyncGenerator<WorkspaceStreamChunk, void, unknown> {
    const response = await client.requestRaw('POST', `/v1/workspace/${slug}/stream-chat`, body);

    if (!response.body) {
        throw new Error('Response body is null — streaming is not supported in this environment');
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let isErrorEvent = false;

    try {
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop() ?? '';

            for (const line of lines) {
                if (line.startsWith('event: error')) {
                    isErrorEvent = true;
                    continue;
                }
                if (line.startsWith('data: ')) {
                    const jsonStr = line.slice(6).trim();
                    if (jsonStr === '[DONE]') {
                        isErrorEvent = false;
                        continue;
                    }
                    let data: WorkspaceStreamChunk;
                    try {
                        data = JSON.parse(jsonStr) as WorkspaceStreamChunk;
                    } catch {
                        continue;
                    }
                    if (isErrorEvent) {
                        throw new Error(data.error ?? 'Unknown streaming error from server');
                    }
                    yield data;
                }
            }
        }
    } finally {
        reader.releaseLock();
    }
}
